pkg/tools/etcd_helper.go

```diff
@@ -116,19 +120,51 @@ func (h *EtcdHelper) decodeNodeList(nodes []*etcd.Node, slicePtr interface{}) er
 			}
 			continue
 		}
-		obj := reflect.New(v.Type().Elem())
-		if err := h.Codec.DecodeInto([]byte(node.Value), obj.Interface().(runtime.Object)); err != nil {
-			return err
-		}
-		if h.Versioner != nil {
-			// being unable to set the version does not prevent the object from being extracted
-			_ = h.Versioner.UpdateObject(obj.Interface().(runtime.Object), node)
+		if obj, found := h.GetFromCache(node.ModifiedIndex); found {
+			v.Set(reflect.Append(v, obj.Elem()))
+		} else {
+			obj := reflect.New(v.Type().Elem())
+			if err := h.Codec.DecodeInto([]byte(node.Value), obj.Interface().(runtime.Object)); err != nil {
+				return err
+			}
+			if h.Versioner != nil {
+				// being unable to set the version does not prevent the object from being extracted
+				_ = h.Versioner.UpdateObject(obj.Interface().(runtime.Object), node)
+			}
+			v.Set(reflect.Append(v, obj.Elem()))
+			if node.ModifiedIndex != 0 {
+				h.AddToCache(node.ModifiedIndex, obj)
+			}
 		}
-		v.Set(reflect.Append(v, obj.Elem()))
 	}
 	return nil
 }
 
+type EtcdCache interface {
+	GetFromCache(index uint64) (obj reflect.Value, ok bool)
+	AddToCache(index uint64, obj reflect.Value)
+}
+
+func (h *EtcdHelper) GetFromCache(index uint64) (obj reflect.Value, ok bool) {
+	h.mutex.RLock()
+	defer h.mutex.RUnlock()
+	obj, ok = h.cache[index]
+	return
+}
+
+func (h *EtcdHelper) AddToCache(index uint64, obj reflect.Value) {
+	h.mutex.Lock()
+	defer h.mutex.Unlock()
+	h.cache[index] = obj
+	if len(h.cache) > 100000 {
```

Comment:
It doesn't have to be modifiable via constructor now.  Exposing it as a public member which can be changed is fine.

> On Apr 27, 2015, at 6:33 AM, Filip Grzadkowski <notifications@github.com> wrote:
> 
> In pkg/tools/etcd_helper.go:
> 
> > +	GetFromCache(index uint64) (obj reflect.Value, ok bool)
> > +	AddToCache(index uint64, obj reflect.Value)
> > +}
> > +
> > +func (h *EtcdHelper) GetFromCache(index uint64) (obj reflect.Value, ok bool) {
> > +	h.mutex.RLock()
> > +	defer h.mutex.RUnlock()
> > +	obj, ok = h.cache[index]
> > +	return
> > +}
> > +
> > +func (h *EtcdHelper) AddToCache(index uint64, obj reflect.Value) {
> > +	h.mutex.Lock()
> > +	defer h.mutex.Unlock()
> > +	h.cache[index] = obj
> > +	if len(h.cache) > 100000 {
> Changing this to be a field in EtcdHelper would require to support modification to that field. I'd prefer to change this to const.
> 
> —
> Reply to this email directly or view it on GitHub.
>